### PR TITLE
AIPO-427: Improve merge conflicts error message

### PR DESCRIPTION
**Pull request by alexlatchford**
To do:
- [x] Add in handling for `MergeConflictException` to nicely format the output of a merge conflict operation.
- [x] Extend CLI tests to cover this scenario.

Before:

```
 /tmp/pricing-triage   master  battenberg upgrade --checkout v2.0.0-alpha.2 --merge-target archetype-update-v2.0.0 --no-input
Removing example code starting
Removing example code complete
Traceback (most recent call last):
  File "/Users/alexla/Code/battenberg/battenberg/cli.py", line 104, in upgrade
    battenberg.upgrade(**kwargs)
  File "/Users/alexla/Code/battenberg/battenberg/core.py", line 291, in upgrade
    self._merge_template_branch(f'Upgraded template \'{template}\'', merge_target)
  File "/Users/alexla/Code/battenberg/battenberg/core.py", line 116, in _merge_template_branch
    f'Cannot merge the template commit ({branch.target}) with the current HEAD '
battenberg.errors.MergeConflictException: Cannot merge the template commit (45657f359bbb40747b38471953e1215b37d36d98) with the current HEAD (<_pygit2.Reference object at 0x106949f90>). Please resolve them manually and run 'git commit' to merge

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.6/bin/battenberg", line 11, in <module>
    load_entry_point('battenberg', 'console_scripts', 'battenberg')()
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/click/decorators.py", line 17, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/Users/alexla/Code/battenberg/battenberg/cli.py", line 106, in upgrade
    raise click.ClickException from error
TypeError: __init__() missing 1 required positional argument: 'message'
```

After:

```
 /tmp/pricing-triage   master  battenberg upgrade --checkout v2.0.0-alpha.2 --merge-target archetype-update-v2.0.0 --no-input
Removing example code starting
Removing example code complete
ignoring NAME extension
On branch archetype-update-v2.0.0
You have unmerged paths.
  (fix conflicts and run "git commit")
  (use "git merge --abort" to abort the merge)

Changes to be committed:

	modified:   Dockerfile
	new file:   pricing_triage/variants/baseline/package.py

Unmerged paths:
  (use "git add/rm <file>..." as appropriate to mark resolution)

	both modified:   .cookiecutter.json
	both modified:   .gitlab-ci.yml
	both modified:   build.gradle
	both modified:   config/hyperparameters.json
	deleted by us:   dags/dag_pricing_triage.yaml
	added by us:     dags/dag_pricing_triage_gitlab.yaml
	both modified:   helm/values.yaml
	both added:      pricing_triage/variants/baseline/validation.py
	both deleted:    pricing_triage/variants/baseline_2/validation.py
	both modified:   versions.yaml

Untracked files:
  (use "git add <file>..." to include in what will be committed)

	pricing_triage/variants/baseline/__init__.py~HEAD
	pricing_triage/variants/baseline/__init__.py~e5920091acd11311045210a05cdcf221bb4ff8c5
        ...


Cannot merge upgrade automatically, please manually resolve the conflicts
```

### CODE REVIEW

File: battenberg/cli.py [hunk 1]
```diff
@@ -102,5 +103,12 @@ def upgrade(ctx, **kwargs):
     try:
         battenberg = Battenberg(open_repository(ctx.obj['target']))
         battenberg.upgrade(**kwargs)
+    except MergeConflictException:
+        # Just run "git status" in a subprocess so we don't have to re-implement the formatting
+        # logic atop pygit2.
+        completed_process = subprocess.run(['git', 'status'], stdout=subprocess.PIPE,
+                                           stderr=subprocess.STDOUT)
+        print(completed_process.stdout.decode('utf-8'))
```
Comment (almonteb) [inline battenberg/cli.py]: should this use the logger/click instead?

File: battenberg/cli.py [hunk 2]
```diff
@@ -102,5 +103,12 @@ def upgrade(ctx, **kwargs):
     try:
         battenberg = Battenberg(open_repository(ctx.obj['target']))
         battenberg.upgrade(**kwargs)
+    except MergeConflictException:
```
Comment (almonteb) [inline battenberg/cli.py]: thoughts on a test around this? should it exit w/ a non-0 exit code?

File: battenberg/cli.py (hunk 1 above)
Comment (alexlatchford) [reply battenberg/cli.py]: Always forget about `click.echo`, fixed in the next commit.

File: battenberg/cli.py (hunk 2 above)
Comment (alexlatchford) [reply battenberg/cli.py]: Yeah I was being lazy by not adding a test, didn't look I bothered writing any tests for the CLI in the last round of development. I'll add some now :) 

I went back and forth about whether this is an error scenario or not. Typically this error will arise if the template changes in such a way that you expect to generate conflicts (ie. when making breaking changes) so in some ways it's a known state and should exit 0. I'm going to put in the exit 1 status though as I think if the `battenberg upgrade` command was ever chained then the next function likely wouldn't want to automatically execute as the only way to resolve this state is manually.

Comment (alexlatchford) [reply battenberg/cli.py]: Done in https://github.com/zillow/battenberg/pull/13/commits/d2891c2608c7d3b7e6f92e075aef9820534a7c07

**Pull request merged**